### PR TITLE
[RLlib] SARL trainer and RLModule

### PR DESCRIPTION
**Pull request by avnishn**
Introduces the torch single agent rl trainer and RL module class with tests :) 

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/core/rl_module.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
+class RLModule:
```
Comment (gjoliver) [inline rllib/core/rl_module.py:1]: should @kouroshHakha introduce this API first? :)

File: rllib/core/torch/torch_rl_module.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+from ray.rllib.core.rl_module import RLModule
+
+
+class TorchRLModule(torch.nn.Module, RLModule):
+    def __init__(self, config):
+        super().__init__()
+        self.config = config
+
+    def forward(self, batch, **kwargs):
```
Comment (gjoliver) [inline rllib/core/torch/torch_rl_module.py:11]: RLModule doesn't have forward() though? only __call__, forward_train, and forward_inference.

**Review by gjoliver — COMMENTED** [review]
let's talk about how we are going to get this work in in person :)

File: rllib/core/torch/torch_rl_module.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import torch
+
+from ray.rllib.core.rl_module import RLModule
+
+
+class TorchRLModule(torch.nn.Module, RLModule):
```
Comment (smorad) [inline rllib/core/torch/torch_rl_module.py:6]: This looks a bit different than the original design: https://github.com/ray-project/ray/pull/25682/files#diff-56205bdce1995828472f98db75622f44908b50611063270544c90c15756aefeb

File: rllib/core/sarl_trainer.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        results = next(self.training_iterator)
+        self.curr_weights = results[0]["module_weights"]
+        return results
+
+    @staticmethod
+    def compute_loss(batch, fwd_out, **kwargs):
+        raise NotImplementedError
+
+    @staticmethod
+    def compute_grads_and_apply_if_needed(batch, fwd_out, loss_out, **kwargs):
```
Comment (smorad) [inline rllib/core/sarl_trainer.py:15]: I'm not a fan of functions that __might__ have side effects. Is there a reason we don't have a separate `compute_grads` and `apply_grads` function?

File: rllib/core/torch/torch_sarl_trainer.py [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+    @staticmethod
+    def init_rl_module(module_config):
+        raise NotImplementedError
+
+    @staticmethod
+    def init_optimizer(module, optimizer_config):
+        raise NotImplementedError
+
+    @staticmethod
+    def compute_loss_and_update(module, batch, optimizer):
```
Comment (smorad) [inline rllib/core/torch/torch_sarl_trainer.py:52]: What is the difference between `update` and `compute_grads_and_apply`? Intuitively, I would think `update` updates the model weights, but then what does `compute_...` do?

File: rllib/core/torch/torch_sarl_trainer.py [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+        raise NotImplementedError
+
+    @staticmethod
+    def compute_loss(batch, fwd_out, device, **kwargs):
+        raise NotImplementedError
+
+    @staticmethod
+    def compute_grads_and_apply_if_needed(
+        batch, fwd_out, loss_out, rl_module, optimizer, device, **kwargs
+    ):
```
Comment (smorad) [inline rllib/core/torch/torch_sarl_trainer.py:62]: Why do we need all these arguments? Isn't `loss.backward(); optimizer.step()` enough?

Comment (avnishn) [reply rllib/core/torch/torch_sarl_trainer.py:62]: probably but wanted to give as much flexibility as possible here.

File: rllib/core/torch/torch_sarl_trainer.py (hunk 5 above)
Comment (avnishn) [reply rllib/core/torch/torch_sarl_trainer.py:52]: `train` / `update` are user facing methods that the user calls on a `TorchSARLTrainer` instance. `compute_grads_and_apply` is called by ray train workers

File: rllib/core/torch/torch_rl_module.py (hunk 2 above)
Comment (avnishn) [reply rllib/core/torch/torch_rl_module.py]: ok so the reason I did this is because I think the DDP wrapper requires that forward is implemented by the module that is being optimized. Although its possible we might be able to get away with just implementing __call__

File: rllib/core/sarl_trainer.py [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    def train(self, batch):
+        # this is the function is called by the user to train the model
+        for queue in self.queues:
+            queue.put(batch)
+        results = next(self.training_iterator)
+        self.curr_weights = results[0]["module_weights"]
+        return results
+
+    @staticmethod
+    def compute_loss(batch, fwd_out, **kwargs):
```
Comment (kouroshHakha) [inline rllib/core/sarl_trainer.py:11]: All these methods are defined here as an abstraction while they are neither an interface api nor are being user directly in this class. It's not clear why so many methods are needed here.  For example how is `compute_loss` gonna be used? should another module call this method directly or should this be called internally when you perform update?

File: rllib/core/torch/torch_sarl_trainer.py [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+from ray.train import TrainingIterator
+from ray.train._internal.checkpoint import CheckpointManager
+from ray.train._internal.dataset_spec import RayDatasetSpec
+
+
+class TorchSARLTrainer(SARLTrainer):
+    def __init__(self, config):
+        num_workers = config.get("num_gpus", 0) or 1
+        use_gpu = bool(config.get("num_gpus", 0))
+        init_rl_module_fn = config.get("rl_module_init_fn", None)
```
Comment (kouroshHakha) [inline rllib/core/torch/torch_sarl_trainer.py:24]: To make faster progress with the design of this module, let's not think about controllability of procedures using config parameters for now. For example `init_rl_module` is a protected method in this class, let's assume if people will do customization by overriding this method in a new class.

File: rllib/core/torch/torch_sarl_trainer.py [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+        """
+        we can't use fractional gpus with this :(
+
+        Leads to nccl errors:
+
+        ncclInvalidUsage: This usually reflects invalid usage of NCCL library
+        (such as too many async ops, too many collectives at once, mixing
+        streams in a group, etc).
+        """
+        scaling_config = ScalingConfig(
```
Comment (kouroshHakha) [inline rllib/core/torch/torch_sarl_trainer.py:35]: none of these functions are framework specific yet. So they can be moved to the parent class. This will solve the confusion on where queues and training_iterator come from and what their behavior should be depends on what `self._make_ray_train_trainer` returns.

File: rllib/core/torch/torch_sarl_trainer.py [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+        rl_module,
+        **kwargs
+    ):
+        raise NotImplementedError
+
+    def get_weights(self):
+        # We need to get the weights from the trainer to the workers
+        return self.curr_weights
+
+    def _make_ray_train_trainer(self, scaling_config, init_rl_module_fn, module_config):
```
Comment (kouroshHakha) [inline rllib/core/torch/torch_sarl_trainer.py:93]: The only torch related function here is `backend_config = TorchConfig()`. so all of this code can live in the base class.

Comment (kouroshHakha) [reply rllib/core/torch/torch_sarl_trainer.py:93]: TorchSARLTrainer only changes the backend_config part of it which can be done by defining an abstract class that needs to be overriden from the base class e.g. `get_backend()`

File: rllib/core/torch/torch_sarl_trainer.py [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+            train_func=train_loop_per_worker,
+            dataset_spec=empty_spec,
+            checkpoint_manager=checkpoint_manager,
+            checkpoint=None,
+            checkpoint_strategy=None,
+        )
+        return queues, training_iterator
+
+    @staticmethod
+    def _training_func(config):
```
Comment (kouroshHakha) [inline rllib/core/torch/torch_sarl_trainer.py:137]: This is an extremely difficult code to understand. There is a lot of information leakage between function calls that makes understanding it very difficult. We don't wanna create another version of what we already have man :)

**Review by kouroshHakha — COMMENTED** [review]
@avnishn Thanks a ton man. It's great that you made a working version of multi-gpu trainer. But this PR needs a good amount of work and re-design iteration until it's ready for production and aligned with the original design. I left some comments here and there for the immediate things that came to my mind. But more fundamentally, I think you need to take a step back and re-think about the structure of the abstractions and class hierarchies. Basically think about which parts are framework specific, which parts are not, How would a user know what to implement from the base class to be able to extend it and so on. You need to put these design decisions into a design doc so that we can discuss alternatives and whether we are leaking unwanted information to other places of the code-base.  Another question that you should be asking is how we can make this code as simple as possible, and very intuitive and follow. The principle idea is simple, you need to shard the data, run parallel fwd passes, run parallel backward passes and aggregate the result. The implementation should also be similarly simple. Don't think that documentation will solve this problem, you need a better and simpler design with no information leakage between modules. Writing a documentation for this kind of code-base would be very easy.

File: rllib/core/torch/torch_sarl_trainer.py [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+        global_size = session.get_world_size()
+        while 1:
+            whole_batch = queue.get()
+            batch_size = np.ceil(len(whole_batch) / global_size)
+            start = batch_size * session.get_local_rank()
+            end = min(start + batch_size, len(whole_batch))
+            batch = whole_batch[int(start) : int(end)]
+            fwd_out = ddp_rl_module(batch, device=device)
+            loss_dict = compute_loss_fn(batch, fwd_out, device=device)
+            info_dict = compute_grads_and_apply_if_needed_fn(
```
Comment (sven1977) [inline rllib/core/torch/torch_sarl_trainer.py:165]: +1 on Steven's comment to split these up into compute_grads and apply_grads and then call them both here in sequence.

File: rllib/core/torch/torch_sarl_trainer.py (hunk 11 above)
Comment (sven1977) [reply rllib/core/torch/torch_sarl_trainer.py:137]: @kouroshHakha:
```
There is a lot of information leakage between function calls that makes understanding it very difficult.
```
This function only gets called once on each train worker, when training starts. The config arg with magic str-keys in it is not pretty, but enforced by Ray Train design, unfortunately.

```
We don't wanna create another version of what we already have man
```
We also don't want to reinvent the wheel by NOT using Ray Train and doing our own DDP/multi-GPU/etc.. Gotta work with what we have here. The upside of all this is that now we'll have a generic design that allows us to scale (on the learner end!), across all algos and DL-frameworks.
That being said, yes, we should try to simplify (and make readable) this as much as possible.
